File: docs-ingestion/ingest.py
from mixedbread import Mixedbread
import os
from dotenv import load_dotenv
import glob
from pathlib import Path

load_dotenv()

# Initialize mixedbread client
mxbai = Mixedbread(
    api_key=os.getenv("MXBAI_API_KEY"),
)

# Vector store ID for NextJS documentation
VECTOR_STORE_ID = os.getenv("VECTOR_STORE_ID")
DOCS_DIR = "docs"


def upload_text_files():
    """Upload all text files from the docs directory to the vector store"""

    print("🚀 Starting NextJS Documentation Ingestion")
    print(f"📁 Source directory: {DOCS_DIR}")
    print(f"🔗 Vector store ID: {VECTOR_STORE_ID}")
    print("=" * 60)

    # Find all text files
    txt_files = glob.glob(f"{DOCS_DIR}/**/*.txt", recursive=True)

    # Filter out progress files
    txt_files = [f for f in txt_files if not f.endswith("_crawl_progress.json")]

    print(f"📄 Found {len(txt_files)} text files to upload")

    if not txt_files:
        print("❌ No text files found in the docs directory")
        return

    uploaded_count = 0
    failed_count = 0

    for i, file_path in enumerate(txt_files, 1):
        try:
            print(f"\n📤 [{i}/{len(txt_files)}] Uploading: {file_path}")

            # Read the file content to extract metadata from frontmatter
            with open(file_path, "r", encoding="utf-8") as f:
                content = f.read()

            # Extract metadata from frontmatter if present
            metadata = {}

            if content.startswith("---"):
                try:
                    parts = content.split("---", 2)
                    if len(parts) >= 3:
                        frontmatter = parts[1].strip()

                        for line in frontmatter.split("\n"):
                            if ":" in line:
                                key, value = line.split(":", 1)
                                key = key.strip()
                                value = value.strip().strip('"')
                                metadata[key] = value
                except:
                    pass

            # Add file path and full text content to metadata
            metadata["file_path"] = file_path
            metadata["file_name"] = Path(file_path).name
            metadata["text"] = content

            # Upload file directly to vector store using tuple format
            with open(file_path, "rb") as f:
                file_content = f.read()

            # Use tuple format: (filename, file_content, content_type)
            file_tuple = (Path(file_path).name, file_content, "text/plain")

            response = mxbai.vector_stores.files.upload(
                vector_store_id=VECTOR_STORE_ID,
                file=file_tuple,
                metadata=metadata,
            )

            print(response)
            uploaded_count += 1

        except Exception as e:
            print(f"❌ Failed to upload {file_path}: {str(e)}")
            failed_count += 1

    # Final summary
    print("\n" + "=" * 60)
    print("🎉 Ingestion Complete!")
    print(f"✅ Successfully uploaded: {uploaded_count} files")
    print(f"❌ Failed uploads: {failed_count} files")
    print(f"📊 Success rate: {(uploaded_count / len(txt_files) * 100):.1f}%")

    if uploaded_count > 0:
        print(f"\n🔍 You can now search the vector store with ID: {VECTOR_STORE_ID}")


if __name__ == "__main__":
    # Check if API key is set
    if not os.getenv("MXBAI_API_KEY"):
        print("❌ Error: MXBAI_API_KEY environment variable not set")
        print("Please add your mixedbread API key to your .env file")
        exit(1)

    # Check if vector store ID is set
    if not VECTOR_STORE_ID:
        print("❌ Error: VECTOR_STORE_ID environment variable not set")
        print("Please add your vector store ID to your .env file")
        exit(1)

    # Check if docs directory exists
    if not os.path.exists(DOCS_DIR):
        print(f"❌ Error: {DOCS_DIR} directory not found")
        print("Please run the scraper first to generate documentation files")
        exit(1)

    # Upload files
    upload_text_files()

File: scripts/ingest.ts
import * as fs from 'fs';
import * as path from 'path';
import { glob } from 'glob';
import dotenv from 'dotenv';
import { mxbai } from '../lib/mxbai';

dotenv.config();

interface FrontmatterData {
  title?: string;
  path?: string;
  source_url?: string;
  content_length?: number;
  [key: string]: string | number | undefined;
}

interface FileMetadata extends FrontmatterData {
  file_path: string;
  file_name: string;
  text: string;
}

function parseFrontmatter(fileContent: string): { frontmatter: FrontmatterData; content: string } {
  if (!fileContent.startsWith('---')) {
    return { frontmatter: {}, content: fileContent };
  }

  const parts = fileContent.split('---');
  if (parts.length < 3) {
    return { frontmatter: {}, content: fileContent };
  }

  const frontmatterText = parts[1].trim();
  const content = parts.slice(2).join('---').trim();
  
  const frontmatter: FrontmatterData = {};
  const lines = frontmatterText.split('\n');
  
  for (const line of lines) {
    const colonIndex = line.indexOf(':');
    if (colonIndex > 0) {
      const key = line.substring(0, colonIndex).trim();
      let value = line.substring(colonIndex + 1).trim();
      
      if (value.startsWith('"') && value.endsWith('"')) {
        value = value.slice(1, -1);
      }
      
      if (key === 'content_length' && !isNaN(Number(value))) {
        frontmatter[key] = Number(value);
      } else {
        frontmatter[key] = value;
      }
    }
  }

  return { frontmatter, content };
}

async function uploadFiles(): Promise<void> {
  console.log('Starting NextJS Documentation Ingestion');
  
  const VECTOR_STORE_ID = process.env.VECTOR_STORE_ID;
  if (!VECTOR_STORE_ID) {
    throw new Error('Missing vector store ID');
  }

  const mdFiles = await glob('content/**/*.md', { ignore: ['**/node_modules/**'] });
  console.log(`Found ${mdFiles.length} markdown files to upload`);

  for (let i = 0; i < mdFiles.length; i++) {
    const filePath = mdFiles[i];
    
    try {
      console.log(`[${i + 1}/${mdFiles.length}] Processing: ${filePath}`);

      const fileContent = fs.readFileSync(filePath, 'utf-8');
      const { frontmatter } = parseFrontmatter(fileContent);
      const fileName = path.basename(filePath);

      const metadata: FileMetadata = {
        ...frontmatter,
        file_path: filePath,
        file_name: fileName,
        // Save the actual text for potential reranking after search
        text: fileContent,
      };

      const fileBuffer = fs.readFileSync(filePath);
      const file = new File([fileBuffer], fileName, { type: 'text/plain' });

      const response = await mxbai.vectorStores.files.upload(
        VECTOR_STORE_ID,
        file,
        { metadata: metadata }
      );

      console.log(`Uploaded: ${fileName} - ID: ${response.id}`);

    } catch (error) {
      console.error(`Failed to process ${filePath}:`, error);
    }
  }

  console.log('\n' + '='.repeat(60));
  console.log('Ingestion Complete!');
}

async function main(): Promise<void> {
  if (!process.env.MXBAI_API_KEY) {
    console.error('Error: MXBAI_API_KEY environment variable not set');
    console.error('Please add your mixedbread API key to your .env file');
    process.exit(1);
  }

  if (!process.env.VECTOR_STORE_ID) {
    console.error('Error: VECTOR_STORE_ID environment variable not set');
    console.error('Please add your vector store ID to your .env file');
    process.exit(1);
  }

  if (!fs.existsSync('content')) {
    console.error('Error: content directory not found');
    console.error('Please ensure the content directory exists with markdown files');
    process.exit(1);
  }

  await uploadFiles();
}

if (require.main === module) {
  main().catch(console.error);
}
